File: ctci/tmp/chapter_4_trees_graphs/js/4.1-routeBetweenNodes.js
const { Graph } = require('./graph')

/**
 * Searches a graph to see if there is a route between two nodes.
 * @param   {Graph}   graph
 * @param   {number}  start
 * @param   {number}  end
 * @returns {boolean}
 */
const search = (graph, start, end) => {
  if (start === end) return true
  const queue = [start]
  const visited = {}
  visited[start.value] = true
  while (queue.length) {
    let curr = queue.shift()
    if (curr) {
      for (let neighbor of curr.adjacents) {
        if (!visited[neighbor.value]) {
          if (neighbor === end) {
            return true
          } else queue.push(neighbor)
        }
        visited[neighbor.value] = true
      }
    }
  }
  return false
}

const g = new Graph()
const sf = g.addVertex('sf')
const hayward = g.addVertex('hayward')
const sj = g.addVertex('sj')
g.addEdge('oakland', 'sf')
g.addEdge('oakland', 'berkeley')
g.addEdge('sf', 'san bruno')
g.addEdge('oakland', 'san leandro')
g.addEdge('san leandro', 'hayward')
g.addEdge('san bruno', 'hayward')
console.log(g)
/*
Graph {
  nodes:
   { sf: Node { value: 'sf', adjacents: [Array] },
     hayward: Node { value: 'hayward', adjacents: [Array] },
     sj: Node { value: 'sj', adjacents: [] },
     oakland: Node { value: 'oakland', adjacents: [Array] },
     berkeley: Node { value: 'berkeley', adjacents: [Array] },
     'san bruno': Node { value: 'san bruno', adjacents: [Array] },
     'san leandro': Node { value: 'san leandro', adjacents: [Array] } } }
*/
console.log(search(g, sf, hayward)) // true
console.log(search(g, sf, sj)) // false
g.addEdge('sj', 'hayward')
console.log(search(g, sf, sj)) // true
g.removeVertex('sj')
console.log(search(g, sf, sj)) // false

File: ctci/chapter_4_trees_graphs/4.1-route_between_nodes.py
Graph = __import__('graph').Graph
GraphNode = __import__('graph').GraphNode


def search(graph: Graph, start: GraphNode, end: GraphNode) -> bool:
    """
    Searches a graph to see if there is a route between two nodes.
    """
    if start == end:
        return True
    queue = [start]
    visited = {}
    visited[start.value] = True
    while len(queue):
        curr = queue.pop()
        if curr:
            for neighbor in curr.adjacent_list:
                if neighbor.value not in visited:
                    if neighbor == end:
                        return True
                    else:
                        queue.append(neighbor)
                visited[neighbor.value] = True
    return False


if __name__ == "main":
    g = Graph(False)
    sf = g.add_vertex('sf')
    hayward = g.add_vertex('hayward')
    sj = g.add_vertex('sj')
    g.add_edge('oakland', 'sf')
    g.add_edge('oakland', 'berkeley')
    g.add_edge('sf', 'san bruno')
    g.add_edge('oakland', 'san leandro')
    g.add_edge('san leandro', 'hayward')
    g.add_edge('san bruno', 'hayward')
    print(g)
    """
    Graph = {
        directed: False
        nodes: ['sf', 'hayward', 'sj', 'oakland', 'berkeley', 'san bruno', 'san leandro']
    }
    """
    print(search(g, sf, hayward))  # True
    print(search(g, sf, sj))  # False
    g.add_edge('sj', 'hayward')
    print(search(g, sf, sj))  # True
    g.remove_vertex('sj')
    print(search(g, sf, sj))  # False
